PAGE XML renderer / export

JKamlah opened the pull request:
Hi everyone,

I've created a PAGE-XML renderer/export.

It's not just a simple PAGE-XML export, it can also produce a textline polygon instead of a simple bounding box, and it can output up to word level.
The output can be customised with three bool parameters 
```
   tessedit_create_page_xml -> "Write .page.xml PAGE file".
   page_xml_polygon -> "Create PAGE file with polygons instead of box values (default:True)".
   page_xml_level -> "Create PAGE with 0 - line or 1 - word level (default:0)".
```

After installing tesseract you can use the preconfigured settings:  'page' . 
page -> Output page.xml file with polygon and line-level

As word-level is more of a niche requirement, you need to enable it via -c:
-c page_xml_polygon=1 // True polygon or False bounding boxes
-c page_xml_level=1 // 0 line or 1 word level

The output is valid for PAGE XML version 2019-07-15.

If the textlines contains only ltr or rtl characters, the output is correct, but for mixed lines (BiDi) I am not quite sure. 
Can anyone help me here?

egorpugin commented:
Can you send a result file of this code for some example image?

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1076 @@
+#include <unordered_set>
+
+#include <allheaders.h>
+
```
```suggestion
+#if (LIBLEPT_MAJOR_VERSION == 1 && LIBLEPT_MINOR_VERSION >= 83) || LIBLEPT_MAJOR_VERSION > 1
#include <array_internal.h>
#include <pix_internal.h>
#endif

```

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1076 @@
+  return result;
+}
+
+} // namespace tesseract
```
Here a terminating linefeed is missing.

stweil reviewed (changes requested):
The CI builds fail because some include statements are missing.

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1080 @@
+/// Block orientation at the current iterator position.
+///
+static tesseract::Orientation 
+GetBlockTextOrientation(const PageIterator *it) {
```
`GetBlockTextOrientation` is unused. There are also currently lots of compiler warnings for unused variables.

stweil replied:
The file also has several lines with trailing blanks which should be removed.

stweil replied:
`typos` finds several typos: reveresed, informations, usefull, simplies.

stweil commented on README.md:
```diff
@@ -36,7 +36,7 @@ Tesseract has **unicode (UTF-8) support**, and can **recognize [more than 100 la
 Tesseract supports **[various image formats](https://tesseract-ocr.github.io/tessdoc/InputFormats)** including PNG, JPEG and TIFF.
 
-Tesseract supports **various output formats**: plain text, hOCR (HTML), PDF, invisible-text-only PDF, TSV and ALTO.
+Tesseract supports **various output formats**: plain text, hOCR (HTML), PDF, invisible-text-only PDF, TSV and ALTO and PAGE.
```
```suggestion
Tesseract supports **various output formats**: plain text, hOCR (HTML), PDF, invisible-text-only PDF, TSV, ALTO and PAGE.
```

stweil commented on include/tesseract/renderer.h:
```diff
@@ -198,6 +198,23 @@ class TESS_API TessAltoRenderer : public TessResultRenderer {
 };
 
+/**
+ * Renders tesseract output into an page xml text string
```
```suggestion
 * Renders Tesseract output into a PAGE XML text string
```

JKamlah commented:
Thanks for suggestion and fixes @stweil 

Here are examples for [eurotext](https://github.com/tesseract-ocr/test/blob/2761899921c08014cf9dbf3b63592237fb9e6ecb/testing/eurotext.tif) and [hebrew](https://github.com/tesseract-ocr/test/blob/2761899921c08014cf9dbf3b63592237fb9e6ecb/testing/hebrew.png) with 'page' and 'page-poly':
https://digi.bib.uni-mannheim.de/~jkamlah/PAGERenderer/

stweil commented:
> Can you send a result file of this code for some example image?

PAGE XML with box coordinates:
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<PcGts xmlns="http://schema.primaresearch.org/PAGE/gts/pagecontent/2019-07-15" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://schema.primaresearch.org/PAGE/gts/pagecontent/2019-07-15 http://schema.primaresearch.org/PAGE/gts/pagecontent/2019-07-15/pagecontent.xsd">
        <Metadata>
                <Creator>Tesseract - 5.3.4-27-ge2f3</Creator>
                <Created>2024-03-20T14:16:51</Created>
                <LastChange>2024-03-20T14:16:51</LastChange>
        </Metadata>
        <Page imageFilename="test/testing/HelloGoogle.tif" imageWidth="400" imageHeight="200" type="content">
                <ReadingOrder>
                        <OrderedGroup id="ro6814328771458624830" caption="Regions reading order">
                                <RegionRefIndexed index="0" regionRef="r0"/>
                        </OrderedGroup>
                </ReadingOrder>
                <TextRegion id="r0" custom="readingOrder {index:0;} readingDirection {left-to-right;} orientation {0;}">
                        <Coords points="10,20 331,20 331,71 10,71"/>
                        <TextLine id="r0l0" readingDirection="left-to-right" custom="readingOrder {index:0;}">
                                <Coords points="10,20 331,20 331,71 10,71"/>
                                <Baseline points="10,59 331,63"/>
                                <TextEquiv index="1" conf="0.9585">
                                        <Unicode>Hello Google</Unicode>
                                </TextEquiv>
                        </TextLine>
                        <TextEquiv index="1" conf="0.9585">
                                <Unicode>Hello Google</Unicode>
                        </TextEquiv>
                </TextRegion>
                </Page>
</PcGts>
```

PAGE XML with polygon:
```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<PcGts xmlns="http://schema.primaresearch.org/PAGE/gts/pagecontent/2019-07-15" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://schema.primaresearch.org/PAGE/gts/pagecontent/2019-07-15 http://schema.primaresearch.org/PAGE/gts/pagecontent/2019-07-15/pagecontent.xsd">
        <Metadata>
                <Creator>Tesseract - 5.3.4-27-ge2f3</Creator>
                <Created>2024-03-20T14:19:36</Created>
                <LastChange>2024-03-20T14:19:36</LastChange>
        </Metadata>
        <Page imageFilename="test/testing/HelloGoogle.tif" imageWidth="400" imageHeight="200" type="content">
                <ReadingOrder>
                        <OrderedGroup id="ro6814328771458624830" caption="Regions reading order">
                                <RegionRefIndexed index="0" regionRef="r0"/>
                        </OrderedGroup>
                </ReadingOrder>
                <TextRegion id="r0" custom="readingOrder {index:0;} readingDirection {left-to-right;} orientation {0;}">
                        <Coords points="10,20 331,20 331,71 10,71"/>
                        <TextLine id="r0l0" readingDirection="left-to-right" custom="readingOrder {index:0;}">
                                <Coords points="10,21 42,21 46,30 74,30 79,20 102,20 105,30 132,30 158,22 193,22 197,32 276,32 276,22 298,22 331,23 331,63 298,63 298,71 259,71 256,62 197,62 158,61 74,60 46,60 10,58"/>
                                <Baseline points="10,59 158,60 331,63"/>
                                <TextEquiv index="1" conf="0.9585">
                                        <Unicode>Hello Google</Unicode>
                                </TextEquiv>
                        </TextLine>
                        <TextEquiv index="1" conf="0.9585">
                                <Unicode>Hello Google</Unicode>
                        </TextEquiv>
                </TextRegion>
                </Page>
</PcGts>
```

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1080 @@
+      block_conf = ((res_it-> Confidence(RIL_BLOCK))/100.);
+      page_str << "\t\t<TextRegion id=\"r" << rcnt << "\" "
+      << "custom=\""<< "readingOrder {index:"<< rcnt <<";} "
+      << "readingDirection {"<< WritingDirectionToStr(writing_direction_block)<<";} "
```
The `readingDirection` is not required here (and below) for LTR scripts. I'd prefer to write it conditionally only for RTL scripts because it typically reduces the size of the PAGE XML output significantly.

JKamlah replied:
Agree.

JKamlah commented:
Added Terminating linefeed, removed trailing whitespaces and fixed typos (found by typos-cli).
For the unused functions and variables I need some time.

stweil commented on CMakeLists.txt:
```diff
@@ -784,6 +785,8 @@ set(TESSERACT_CONFIGS
     tessdata/configs/lstmdebug
     tessdata/configs/makebox
+    tessdata/configs/page
+    tessdata/configs/page_poly
```
I suggest to install only `page` and modify it to use polygons instead of bounding boxes.

stweil commented on include/tesseract/baseapi.h:
```diff
@@ -559,6 +559,18 @@ class TESS_API TessBaseAPI {
+   * data structures.
+   */
+  char *GetPAGEText(int page_number);
+
```
Would it be better to combine the two functions in a single function `char *GetPAGEText(int page_number, ETEXT_DESC *monitor = nullptr)`?

JKamlah replied:
For now, it's consistent with all the other renders. See lines above.

stweil replied:
I know, but we are free to improve things if that's possible. I never likes the old API very much. Adding lots of new functions for each new renderer is not very elegant. Maybe someone has a better idea.

stweil commented on src/ccmain/tesseractclass.cpp:
```diff
@@ -340,6 +340,9 @@ Tesseract::Tesseract()
     , BOOL_MEMBER(tessedit_create_hocr, false, "Write .html hOCR output file", this->params())
     , BOOL_MEMBER(tessedit_create_alto, false, "Write .xml ALTO file", this->params())
+    , BOOL_MEMBER(tessedit_create_page, false, "Write .page.xml PAGE file", this->params())
+    , BOOL_MEMBER(tessedit_create_page_polygon, true, "Create the PAGE file with polygons instead of bbox values", this->params())
```
```suggestion
    , BOOL_MEMBER(tessedit_create_page_polygon, true, "Create the PAGE file with polygons instead of box values", this->params())
```

stweil replied:
`bbox` is used internally in the Tesseract code, but not used in the documentation, so we should avoid that name.

stweil commented on src/ccmain/tesseractclass.h:
```diff
@@ -897,6 +897,9 @@ class TESS_API Tesseract : public Wordrec {
   BOOL_VAR_H(tessedit_create_hocr);
   BOOL_VAR_H(tessedit_create_alto);
+  BOOL_VAR_H(tessedit_create_page);
+  BOOL_VAR_H(tessedit_create_page_polygon);
+  BOOL_VAR_H(tessedit_create_page_wordlevel);
```
I suggest to omit the prefix `tessedit_create` and simply use `page_xml_polygon` and `page_xml_wordlevel` (or `page_xml_level` with 0=line, 1=word which might allow a future implementation to add 2=glyph).

For hocr there exist already the parameters `hocr_font_info` and `hocr_char_boxes`, so the new parameters would fit better into that schema.

stweil commented on src/ccmain/tesseractclass.h:
```diff
@@ -897,6 +897,9 @@ class TESS_API Tesseract : public Wordrec {
   BOOL_VAR_H(tessedit_create_txt);
   BOOL_VAR_H(tessedit_create_hocr);
   BOOL_VAR_H(tessedit_create_alto);
+  BOOL_VAR_H(tessedit_create_page);
```
_page_ is already used in a different context for several parameters, so maybe `tessedit_create_page_xml` would be clearer.

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1132 @@
+      continue;
+    }
+
+    int left, top, right, bottom;
```
```suggestion
```

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1132 @@
+        x0 = x_min;
+      }
+    } else if (x0 > x_max) {
+      if (ptaGetCount(baseline_recalc_pts) > 0) {
```
`baseline_recalc_pts` is uninitialized when used here.

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1132 @@
+  numaGetMedian(poly_bl_delta, &delta_median);
+  numaGetRankValue(poly_bl_delta, 0.25, NULL, 0, &delta_median_Q1);
+  numaGetRankValue(poly_bl_delta, 0.75, NULL, 0, &delta_median_Q3);
+  delta_median_IQR = abs(delta_median_Q3 - delta_median_Q1);
```
`delta_median_IQR` is set, but never used.

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1132 @@
+  int num_pts, num_bin, index = 0, p_index, offset, x, y, x0, y0, x1, y1;
+  float m, b;
+  float x_min, y_min, x_max, y_max;
+  float x_min_bl, y_min_bl, x_max_bl, y_max_bl;
```
```suggestion
  float x_min_bl, y_min_bl;
```

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1132 @@
+  }
+
+  num_pts = ptaGetCount(baseline_pts);
+  index = 0;
```
```suggestion
```

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1132 @@
+///
+Pta *FitBaselineIntoLinePolygon(Pta *bottom_pts, Pta *baseline_pts,
+                                tesseract::WritingDirection writing_direction) {
+  int num_pts, num_bin, index = 0, p_index, offset, x, y, x0, y0, x1, y1;
```
```suggestion
  int num_pts, num_bin, x0, y0, x1, y1;
```

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1132 @@
+Pta *ClipAndSimplifyBaseline(Pta *bottom_pts, Pta *baseline_pts,
+                             tesseract::WritingDirection writing_direction) {
+  int num_pts;
+  float m, b, x, y, x0, y0, x1, y1;
```
```suggestion
  float m, b, x0, y0, x1, y1;
```

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1132 @@
+/// Create a hull for overlapping areas
+///
+Pta *RecalcPolygonline(Pta *pts, bool upper) {
+  int num_pts, num_bin, index = 0, p_index, offset;
+  int m, b, x, y, x0, y0, x1, y1;
```
```suggestion
  int num_pts, num_bin, index = 0;
  int y, x0, y0, x1, y1;
```

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1132 @@
+/// currently not necessary)
+///
+static void SimplifyLinePolygon(Pta *polyline, int tolerance, bool upper) {
+  int num_pts, x0, y0, x1, y1, x2, y2, x3, y3, index = 1;
```
```suggestion
  int x0, y0, x1, y1, x2, y2, x3, y3, index = 1;
```

stweil commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1132 @@
+static void AddPointToWordPolygon(
+    const ResultIterator *res_it, PageIteratorLevel level, Pta *word_top_pts,
+    Pta *word_bottom_pts, tesseract::WritingDirection writing_direction) {
+  int num_pts, left, top, right, bottom;
```
```suggestion
  int left, top, right, bottom;
```

stweil commented:
Maybe we can merge this new feature in a new 5.4.0 pre-release.

JKamlah commented:
Sure. I would wait until the end of the week for final comments and then do a squash commit and pull request.

egorpugin commented:
Should we squash long PRs?

JKamlah commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1154 @@
+/// Sort baseline points ascending and deleting duplicates
+///
+Pta *SortBaseline(Pta *baseline_pts,
+                  tesseract::WritingDirection writing_direction) {
```
This input var can be removed. Initially, the function depended on the writing direction, but this part has been removed.

JKamlah commented on src/api/pagerenderer.cpp:
```diff
@@ -0,0 +1,1154 @@
+///
+Pta *RecalcPolygonline(Pta *pts, bool upper) {
+  int num_pts, num_bin, index = 0;
+  int y, x0, y0, x1, y1;
```
Thanks a lot @stweil
